### PR TITLE
Simplify Message Serialization and Parse TLV Suffix

### PR DESCRIPTION
**Pull request by TheBlueMatt**
See individual commits

### CODE REVIEW

File: lightning/src/ln/msgs.rs [hunk 1]
```diff
@@ -1213,47 +1193,51 @@ impl_writeable!(OpenChannel, {
...
-impl_writeable!(UpdateFulfillHTLC, {
+impl_writeable_msg!(UpdateFulfillHTLC, {
 	channel_id,
 	htlc_id,
 	payment_preimage
-});
+}, {});
 
+// Note that this is written as a part of ChannelManager objects, and thus cannot change its
+// serialization format in a non-backwards-compatible way.
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:1234]: Should this say "cannot change its serialization format in a *backwards-compatible* way"? And below

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -4841,10 +4841,74 @@ impl_writeable_tlv_based!(PendingHTLCInfo, {
...
+				let res = Readable::read(&mut s)?;
+				s.eat_remaining()?; // Return ShortRead if there's actually not enough bytes
+				Ok(HTLCFailureMsg::Relay(res))
+			},
+			3 => {
+				let length: BigSize = Readable::read(reader)?;
+				let mut s = FixedLengthReader::new(reader, length.0);
+				let res = Readable::read(&mut s)?;
+				s.eat_remaining()?; // Return ShortRead if there's actually not enough bytes
+				Ok(HTLCFailureMsg::Relay(res))
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:4905]: I think this one should be `::Malformed`

**Review by valentinewallace — COMMENTED** [review]
Nice cleanup! Minor comments, otherwise ACK

File: lightning/src/ln/msgs.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs:1234]: Well if the change is backwards-compatible then presumably its ok? I'm not actually sure *how* you'd make a backwards compatible change here, but I suppose its theoretically possible?

File: lightning/src/ln/channelmanager.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:4905]: Eek!

File: lightning/src/ln/msgs.rs (hunk 1 above)
Comment (valentinewallace) [reply lightning/src/ln/msgs.rs:1234]: I see. I'm just taking issue with the double negative ("cannot change in a non-backwards-compatible way") -- maybe rephrase to "can only be changed in a backwards-compatible way"

File: lightning/src/ln/msgs.rs [hunk 3]
```diff
@@ -1225,56 +1193,55 @@ impl_writeable_len_match!(OpenChannel, {
...
 	payment_preimage
-});
+}, {});
 
-impl_writeable_len_match!(OnionErrorPacket, {
-		{ OnionErrorPacket { ref data, .. }, 2 + data.len() }
-	}, {
+// Note that this is written as a part of ChannelManager objects, and thus cannot change its
+// serialization format in a way which assumes we know the total serialized length/message end
+// position..
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:1235]: nit: s/../. and below

File: lightning/src/ln/msgs.rs [hunk 4]
```diff
@@ -1499,10 +1461,7 @@ impl Readable for UnsignedChannelAnnouncement {
 	}
 }
 
-impl_writeable_len_match!(ChannelAnnouncement, {
-		{ ChannelAnnouncement { contents: UnsignedChannelAnnouncement {ref features, ref excess_data, ..}, .. },
-			2 + 32 + 8 + 4*33 + features.byte_count() + excess_data.len() + 4*64 }
-	}, {
+impl_writeable!(ChannelAnnouncement, {
```
Comment (valentinewallace) [inline lightning/src/ln/msgs.rs:1464]: Why not `impl_writeable_msg`?

**Review by valentinewallace — APPROVED** [review]

File: lightning/src/ln/msgs.rs (hunk 4 above)
Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs:1464]: It doesn't make sense, I think - the contained UnsignedChannelAnnouncement always reads until the end, so there can never be bytes left after reading it that fit in a TLV. Any TLV entries will be in the unsigned one.

File: lightning/src/util/ser.rs [hunk 5]
```diff
@@ -44,18 +44,13 @@ pub const MAX_BUF_SIZE: usize = 64 * 1024;
 pub trait Writer {
 	/// Writes the given buf out. See std::io::Write::write_all for more
 	fn write_all(&mut self, buf: &[u8]) -> Result<(), io::Error>;
-	/// Hints that data of the given size is about the be written. This may not always be called
-	/// prior to data being written and may be safely ignored.
-	fn size_hint(&mut self, size: usize);
```
Comment (jkczyz) [inline lightning/src/util/ser.rs:49]: Update `Writer` docs now that `size_hint` has been removed.

**Review by jkczyz — APPROVED** [review]
FWIW, commit 27c6ee2e97ab0fe61495e05104703106c8ce503e refers to "previous commit" but the commit in question is a few commits back.

**Comment by TheBlueMatt** [comment]
Reordered the commits a bit and updated writer docs. Note that at this point we should probably remove writer, but I'll leave that as a followup huge-diff change.

**Review by valentinewallace — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
Squashed with no further changes, will land after CI.

```
$ git diff-tree -U1 41cf0c2e 997dc5f5
$
```
